File: android/walletsdk/src/main/java/com/coinbase/android/nativesdk/CoinbaseWalletSDK.kt
package com.coinbase.android.nativesdk

import android.content.Context
import android.content.Intent
import android.net.Uri
import com.coinbase.android.nativesdk.key.IKeyManager
import com.coinbase.android.nativesdk.key.KeyManager
import com.coinbase.android.nativesdk.message.JSON
import com.coinbase.android.nativesdk.message.request.Account
import com.coinbase.android.nativesdk.message.request.Action
import com.coinbase.android.nativesdk.message.request.ETH_REQUEST_ACCOUNTS
import com.coinbase.android.nativesdk.message.request.RequestContent
import com.coinbase.android.nativesdk.message.request.RequestConverter
import com.coinbase.android.nativesdk.message.request.UnencryptedRequestContent
import com.coinbase.android.nativesdk.message.request.UnencryptedRequestMessage
import com.coinbase.android.nativesdk.message.request.unsupportedHandshakeActions
import com.coinbase.android.nativesdk.message.response.ActionResult
import com.coinbase.android.nativesdk.message.response.FailureResponseCallback
import com.coinbase.android.nativesdk.message.response.ResponseConverter
import com.coinbase.android.nativesdk.message.response.ResponseHandler
import com.coinbase.android.nativesdk.message.response.ResponseResult
import com.coinbase.android.nativesdk.message.response.SuccessHandshakeResponseCallback
import com.coinbase.android.nativesdk.message.response.SuccessRequestResponseCallback
import com.coinbase.android.nativesdk.task.TaskManager
import kotlinx.serialization.decodeFromString
import java.security.interfaces.ECPublicKey
import java.util.Date
import java.util.UUID

class CoinbaseWalletSDK internal constructor(
    private val hostPackageName: String,
    private val scheme: String,
    private val keyManager: IKeyManager
) {


    val isConnected: Boolean get() = keyManager.peerPublicKey != null
    val ownPublicKey: ECPublicKey get() = keyManager.ownPublicKey
    val peerPublicKey: ECPublicKey? get() = keyManager.peerPublicKey

    private val config: ClientConfiguration.Configuration
        get() = ClientConfiguration.config


    private val launchWalletIntent: Intent?
        get() = config.context.packageManager.getLaunchIntentForPackage(hostPackageName)

    val isCoinbaseWalletInstalled get() = launchWalletIntent != null

    init {
        instances[scheme] = this
    }

    constructor(
        hostPackageName: String,
        scheme: String,
    ) : this(
        hostPackageName,
        scheme,
        KeyManager(ClientConfiguration.config.context, hostPackageName),
    )

    /**
     * Make handshake request to get session key from wallet
     * @param initialActions Batch of actions that you'd want to execute after successful handshake. `eth_requestAccounts` by default.
     * @param onResponse Response callback with regular response result and optional parsed [Account] object.
     */
    fun initiateHandshake(
        initialActions: List<Action>? = null,
        onResponse: (ResponseResult, Account?) -> Unit
    ) {
        resetSession()

        val hasIllegalAction = initialActions?.any { unsupportedHandshakeActions.contains(it.method) } == true
        if (hasIllegalAction) {
            onResponse(Result.failure(CoinbaseWalletSDKError.InvalidHandshakeRequest), null)
            return
        }

        val message = UnencryptedRequestMessage(
            uuid = UUID.randomUUID().toString(),
            version = sdkVersion,
            timestamp = Date(),
            sender = keyManager.ownPublicKey,
            content = UnencryptedRequestContent(
                handshake = RequestContent.Handshake(
                    appId = config.context.packageName,
                    callback = config.domain.toString(),
                    appName = config.name.orEmpty(),
                    appIconUrl = config.iconUrl,
                    initialActions = initialActions
                )
            ),
            callbackUrl = config.domain.toString()
        )

        send(message) { result ->
            // Get index of eth_requestAccounts action
            val requestAccountsIndex = initialActions?.indexOfFirst { it.method == ETH_REQUEST_ACCOUNTS } ?: -1
            if (requestAccountsIndex == -1) {
                onResponse(result, null)
                return@send
            }

            // Get response from Wallet at index
            val requestAccountsResult = result.getOrNull()?.getOrNull(requestAccountsIndex)
            if (requestAccountsResult !is ActionResult.Result) {
                onResponse(result, null)
                return@send
            }

            val account = runCatching {
                JSON.decodeFromString<Account>(requestAccountsResult.value)
            }.getOrNull()

            onResponse(result, account)
        }
    }

    fun initiateHandshake(
        initialActions: List<Action>? = null,
        onSuccess: SuccessHandshakeResponseCallback,
        onFailure: FailureResponseCallback
    ) {
        initiateHandshake(initialActions) { result, account ->
            result
                .onSuccess { onSuccess.call(it, account) }
                .onFailure { onFailure.call(it) }
        }
    }

    /**
     * Make regular requests. It requires session key you get after successful handshake.
     */
    fun makeRequest(
        request: RequestContent.Request,
        onResponse: ResponseHandler
    ) {
        val message = UnencryptedRequestMessage(
            uuid = UUID.randomUUID().toString(),
            version = sdkVersion,
            timestamp = Date(),
            sender = keyManager.ownPublicKey,
            content = UnencryptedRequestContent(request = request),
            callbackUrl = config.domain.toString()
        )

        send(message, onResponse)
    }

    fun makeRequest(
        request: RequestContent.Request,
        onSuccess: SuccessRequestResponseCallback,
        onFailure: FailureResponseCallback
    ) {
        makeRequest(request) { result ->
            result
                .onSuccess { onSuccess.call(it) }
                .onFailure { onFailure.call(it) }
        }
    }

    /**
     * Handle incoming deep links
     * @param url deep link url
     * @return `false` if the input was not response message type, or `true` if SDK handled the input
     */
    fun handleResponse(url: Uri): Boolean {
        if (!isWalletSegueResponseURL(url)) {
            return false
        }

        val ownPublicKey = keyManager.ownPublicKey
        val peerPublicKey = keyManager.peerPublicKey

        val message = ResponseConverter.decode(
            url = url,
            ownPublicKey = ownPublicKey,
            ownPrivateKey = keyManager.ownPrivateKey,
            peerPublicKey = peerPublicKey
        )

        if (peerPublicKey == null && message.sender != ownPublicKey) {
            keyManager.storePeerPublicKey(message.sender as ECPublicKey)
        }

        return TaskManager.handleResponse(message)
    }

    fun resetSession() {
        TaskManager.reset(scheme)
        keyManager.resetKeys()
    }

    private fun send(message: UnencryptedRequestMessage, onResponse: ResponseHandler) {
        val uri: Uri
        try {
            uri = RequestConverter.encode(
                message = message,
                recipient = Uri.parse(scheme),
                ownPrivateKey = keyManager.ownPrivateKey,
                peerPublicKey = keyManager.peerPublicKey
            )
        } catch (e: Throwable) {
            when (e) {
                is CoinbaseWalletSDKError -> onResponse(Result.failure(e))
                else -> onResponse(Result.failure(CoinbaseWalletSDKError.EncodingFailed))
            }
            return
        }

        val intent = launchWalletIntent
        if (intent == null) {
            onResponse(Result.failure(CoinbaseWalletSDKError.OpenWalletFailed))
            return
        }

        // Prevent intent from launching app in new window
        intent.type = Intent.ACTION_VIEW
        if (intent.flags and Intent.FLAG_ACTIVITY_NEW_TASK > 0) {
            intent.flags = intent.flags and Intent.FLAG_ACTIVITY_NEW_TASK.inv()
        }

        intent.data = uri

        TaskManager.registerResponseHandler(message, onResponse, scheme)
        openIntent?.invoke(intent)
    }

    private fun isWalletSegueResponseURL(uri: Uri): Boolean {
        return uri.host == config.domain.host && uri.path == config.domain.path && uri.getQueryParameter("p") != null
    }

    companion object {
        private val instances: MutableMap<String, CoinbaseWalletSDK> = mutableMapOf()

        var openIntent: ((Intent) -> Unit)? = null

        @JvmStatic
        var sdkVersion = BuildConfig.LIBRARY_VERSION_NAME
            private set

        @JvmStatic
        fun configure(domain: Uri, context: Context, appName: String? = null, appIconUrl: String? = null) {
            ClientConfiguration.configure(domain, context, appName, appIconUrl)
        }

        @JvmStatic
        fun setOpenIntentCallback(callback: OpenIntentCallback?) {
            openIntent = if (callback == null) {
                null
            } else {
                { intent -> callback.call(intent) }
            }
        }

        @JvmStatic
        fun getClient(wallet: Wallet): CoinbaseWalletSDK {
            if (openIntent == null) {
                throw IllegalStateException("Must initialize open intent callback before getting CoinbaseWalletSDK instance")
            }

            return instances[wallet.url] ?: CoinbaseWalletSDK(
                hostPackageName = wallet.packageName,
                scheme = wallet.url,
                keyManager = KeyManager(ClientConfiguration.config.context, wallet.packageName)
            )
        }

        @JvmStatic
        fun handleResponseUrl(uri: Uri): Boolean {
            try {
                val content = ResponseConverter.decodeWithoutDecryption(uri).content
                val requestId = when {
                    content.response != null -> checkNotNull(content.response.requestId) { "requestId not found" }
                    content.failure != null -> content.failure.requestId
                    else -> return false
                }

                val task = TaskManager.findTask(requestId) ?: return false
                return instances[task.host]?.handleResponse(uri) ?: false
            } catch (e: IllegalStateException) {
                // Fallback to CB Wallet instance
                return instances[DefaultWallets.coinbaseWallet.url]?.handleResponse(uri) ?: false
            }
        }

        @JvmStatic
        fun appendVersionTag(tag: String) {
            sdkVersion += "/$tag"
        }
    }
}

File: ios/CoinbaseWalletSDK/CoinbaseWalletSDK.swift
//
//  CoinbaseWalletSDK.swift
//  WalletSegue
//
//  Created by Jungho Bang on 5/20/22.
//

import Foundation
import CryptoKit
import UIKit

public final class CoinbaseWalletSDK {
    
    static public func isCoinbaseWalletInstalled() -> Bool {
        return UIApplication.shared.canOpenURL(URL(string: "cbwallet://")!)
    }
    
    static public func getCoinbaseWalletMWPVersion() -> String? {
        if UIApplication.shared.canOpenURL(URL(string: "mwp+1.1://")!) {
            return "1.1"
        } else if isCoinbaseWalletInstalled() {
            return "1.0"
        } else {
            return nil
        }
    }
    
    // MARK: - Constructor
    
    // TODO: revisit singleton approach
    static private var callback: URL?
    static public func configure(callback: URL) {
        guard self.callback == nil else {
            assertionFailure("`CoinbaseWalletSDK.configure` should be called only once.")
            return
        }
        
        if callback.pathComponents.count < 2 { // [] or ["/"]
            self.callback = callback.appendingPathComponent("wsegue")
        } else {
            self.callback = callback
        }
    }
    
    static public var shared: CoinbaseWalletSDK = {
        guard
              let callback = CoinbaseWalletSDK.callback else {
            preconditionFailure("Missing configuration: call `CoinbaseWalletSDK.configure` before accessing the `shared` instance.")
        }
        
        return CoinbaseWalletSDK(callback: callback)
    }()
    
    // MARK: - Properties
    
    private let appId: String
    private var host: URL! // TODO: no optionals
    private let callback: URL
    
    private var keyManager: KeyManager! // TODO: no optionals
    private lazy var taskManager: TaskManager = {
        TaskManager()
    }()
    
    private init(callback: URL) {
        self.callback = callback
        self.appId = Bundle.main.bundleIdentifier!
        self.updateHost(host: URL(string: "https://wallet.coinbase.com/wsegue")!) // TODO: revisit host setup process
    }
    
    public func updateHost(host: URL) {
        self.host = host
        self.keyManager = KeyManager(host: host)
    }
    
    // MARK: - Send message
    
    /// Make handshake request to get session key from wallet
    /// - Parameters:
    ///   - initialActions: Batch of actions that you'd want to execute after successful handshake. `eth_requestAccounts` by default.
    ///   - onResponse: Response callback with regular response result and optional parsed `Account` object.
    public func initiateHandshake(
        initialActions: [Action]? = [Action(jsonRpc: .eth_requestAccounts)],
        onResponse: @escaping (ResponseResult, Account?) -> Void
    ) {
        let hasUnsupportedAction = initialActions?.contains { action in
            ["eth_signTransaction", "eth_sendTransaction"].contains(where: { action.method == $0 })
        }
        
        guard hasUnsupportedAction != true else {
            onResponse(.failure(Error.invalidHandshakeRequest), nil)
            return
        }
        
        try? keyManager.resetOwnPrivateKey()
        let message = RequestMessage(
            uuid: UUID(),
            sender: keyManager.ownPublicKey,
            content: .handshake(
                appId: appId,
                callback: callback,
                initialActions: initialActions
            ),
            version: CoinbaseWalletSDK.version,
            timestamp: Date(),
            callbackUrl: callback.absoluteString
        )
        self.send(message) { result in
            guard
                let requestAccountsIndex = initialActions?.firstIndex(where: { $0.method == "eth_requestAccounts" }),
                let content = try? result.get().content,
                content.indices.contains(requestAccountsIndex),
                case .success(let accountJson) = content[requestAccountsIndex],
                let account = try? accountJson.decode(as: Account.self)
            else {
                onResponse(result, nil)
                return
            }
            
            onResponse(result, account)
        }
    }
    
    /// Make regular requests. It requires session key you get after successful handshake.
    public func makeRequest(_ request: Request, onResponse: @escaping ResponseHandler) {
        let message = RequestMessage(
            uuid: UUID(),
            sender: keyManager.ownPublicKey,
            content: .request(actions: request.actions, account: request.account),
            version: CoinbaseWalletSDK.version,
            timestamp: Date(),
            callbackUrl: callback.absoluteString
        )
        return self.send(message, onResponse)
    }
    
    private func send(_ request: RequestMessage, _ onResponse: @escaping ResponseHandler) {
        let url: URL
        do {
            url = try MessageConverter.encode(request, to: host, with: keyManager.symmetricKey)
        } catch {
            onResponse(.failure(error))
            return
        }
        
        UIApplication.shared.open(url) { result in
            guard result == true else {
                onResponse(.failure(Error.openUrlFailed))
                return
            }
            
            self.taskManager.registerResponseHandler(for: request, onResponse)
        }
    }
    
    // MARK: - Receive message
    
    private func isWalletSegueMessage(_ url: URL) -> Bool {
        return url.host == callback.host && url.path == callback.path
    }

    /// Handle incoming deep links
    /// - Parameter url: deep link url
    /// - Returns: `false` if the input was not response message type, `true` if SDK handled the input, or throws error if it failed to decode response.
    @discardableResult
    public func handleResponse(_ url: URL) throws -> Bool {
        guard isWalletSegueMessage(url) else {
            return false
        }
        
        let response = try decodeResponse(url)
        taskManager.runResponseHandler(with: response)
        return true
    }
    
    private func decodeResponse(_ url: URL) throws -> ResponseMessage {
        if let symmetricKey = keyManager.symmetricKey {
            return try MessageConverter.decode(url, with: symmetricKey)
        }
        
        // no symmetric key yet
        let encryptedResponse: EncryptedResponseMessage = try MessageConverter.decodeWithoutDecryption(url)
        let request = taskManager.findRequest(for: encryptedResponse)
        guard case .handshake = request?.content else {
            throw Error.missingSymmetricKey
        }
        
        try handleHandshakeResponse(encryptedResponse)
        
        return try encryptedResponse.decrypt(with: keyManager.symmetricKey)
    }
    
    // MARK: - Session
    
    public func isConnected() -> Bool {
        return keyManager.symmetricKey != nil
    }
    
    public var ownPublicKey: PublicKey {
        return keyManager.ownPublicKey
    }
    
    public var peerPublicKey: PublicKey? {
        return keyManager.peerPublicKey
    }
    
    @discardableResult
    public func resetSession() -> Result<Void, Swift.Error> {
        do {
            taskManager.reset()
            try keyManager.resetOwnPrivateKey()
            return .success(())
        } catch {
            return .failure(error)
        }
    }
    
    private func handleHandshakeResponse(_ response: EncryptedResponseMessage) throws {
        if case .failure = response.content {
            return
        }
        
        try keyManager.storePeerPublicKey(response.sender)
    }
}
